packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,280 @@
+
+        // Calculate the reward per proposal and per vote
+        t.rewardPerProposal = proposalRewardForPeriod / numEligibleProposals;
+        t.rewardPerVote = votingRewardForPeriod / t.numEligibleVotes;
+
+        //// Second loop over the proposals:
+        //// 1. Skip proposals that were deleted for non eligibility.
+        //// 2. Reward proposal's clientId.
+        //// 3. Reward the clientIds that faciliated voting.
+        //// 4. Make sure all voting clientIds were included.
+
+        for (uint256 i; i < proposals.length; ++i) {
+            // skip non eligible deleted proposals
+            if (proposals[i].endBlock == 0) continue;
+
+            uint32 clientId = proposals[i].clientId;
+            if (clientId != 0) {
+                clientBalances[clientId] += t.rewardPerProposal;
+            }
+
+            uint256 votesInProposal;
+            NounsDAOV3Types.ClientVoteData[] memory voteData = proposals[i].voteData;
+            uint256 votes;
+            for (uint256 j; j < votingClientIds.length; ++j) {
+                clientId = votingClientIds[j];
+                votes = voteData[j].votes;
+                if (clientId != 0) {
+                    clientBalances[clientId] += votes * t.rewardPerVote;
+                }
+                votesInProposal += votes;
+            }
+            require(
+                votesInProposal == proposals[i].forVotes + proposals[i].againstVotes + proposals[i].abstainVotes,
+                'not all votes accounted'
+            );
+        }
+    }
+
+    function withdrawClientBalance(uint32 clientId, uint256 amount, address to) public {
+        require(ownerOf(clientId) == msg.sender, 'must be client NFT owner');
+
+        clientBalances[clientId] -= amount;
+
+        ethToken.transfer(to, amount);
+    }
+
+    struct ProposalRewardsParams {
+        uint32 lastProposalId;
+        uint256 expectedNumEligibleProposals;
+        uint256 expectedNumEligibleVotes;
+        uint256 firstNounId;
+        uint256 lastNounId;
+        uint32[] votingClientIds;
+    }
+
+    function getAuctionRevenueBetweenNouns(
+        uint256 firstNounId,
+        uint256 oneAfterLastNounId,
+        uint256 endTimestamp
+    ) internal view returns (uint256) {
+        INounsAuctionHouseV2.Settlement[] memory s = auctionHouse.getSettlements(
+            firstNounId,
+            oneAfterLastNounId + 1,
+            true
+        );
+        require(s[s.length - 2].blockTimestamp <= endTimestamp, 'second to last auction must be before end ts');
+        require(s[s.length - 1].blockTimestamp > endTimestamp, 'last auction must be after end ts');
+
+        return sumAuctionsExcludingLast(s);
+    }
+
+    function sumAuctionsExcludingLast(INounsAuctionHouseV2.Settlement[] memory s) internal pure returns (uint256 sum) {
+        for (uint256 i = 0; i < s.length - 1; ++i) {
+            sum += s[i].amount;
+        }
+    }
+
+    function registerClient() public returns (uint32) {
```

Comment:
let's remember to add some basic client metadata here
client URL seems like the bare minimum 
a display name could be nice as well

and wdyt about introducing a descriptor external dep to make it safer to render art in the future?